### ISSUE TITLE
CPU usage at 100% when executing (read)

### ISSUE THREAD

**Issue by mnieper** [ashinn/chibi-scheme#295]
CPU usage is at 100% on my machine when (read) is waiting for my input while the following program is being executed:

    (import (scheme base) (scheme read))
    (read)

**Comment by cwds** [comment]
This sounded like an interesting problem so I had a look at it.  I have only been looking through the Chibi-scheme code base for a few days to evaluate it's use for an upcoming project, so I'm no expert on the intent behind the code.

The code as listed certainly does peg a single CPU at 100% when executed as a script; however, when a '(read)' is executed in a REPL, CPU usage is right near 0% as you would expect.

The 100% CPU usage is coming from a busy loop inside the virtual machine.  The (read) compiles down to code using a SEXP_OP_PEEK_CHAR opcode, which is responsible for fetching data from the input port.  The busy loop is inside the sexp_apply() function, located in line 1013 of the vm.c file.

The code starting at the 'loop:' tag first does a few things to run the scheduler (part of the green threads implementation I believe), then enters the "big switch" statement for the current opcode.  The SEXP_OP_PEEK_CHAR opcode is selected.  The code there dutifuly attempts to read from the input port, in this case is STDIN which has been set to non-blocking.  As there is no data there yet but the port is set to non-blocking, the read results in an errno of EAGAIN, which means that the resource is temporarily unavailable.  The SEXP_OP_PEEK_CHAR_CODE then decrements the opcode pointer back to the current opcode, resets the 'fuel' (a counter for the green threading scheduler), then jumps back to the 'loop:' tag to do it all over again.  This busy-loop polling on fgetc() is what is pegging the CPU usage at 100%.

There is code in place to prevent this kind of thing, but it isn't being called in this case.  Up around line 1013 in vm.c, the code:
tmp1 = sep_global(cox, SEXP_G_THREADS_SCHEDULER);
if (sexp_applcablep(tmp1) && sexp_not(sep_global(ctx, SEXP_G_ATOMIC_P))) ... 

If both these conditions are true, sexp_apply1 is called, which in turn calls sep_scheduler in threads.c, line 623, which will call usleep(usecs) which will "take a nap to avoid busy looping".

When using the REPL, SRFI 18 appears to be loaded by default.  SRFI 18 code sets the global SEXP_G_THREADS_SCHEDULER, which in turn allows the if condition on line 1015 of vm.c to be true, allowing the scheduler to be called and the usleep to be run, breaking the busy loop.

When running a script as reported by this error report, SRFI 18 is NOT called by default, which means SEXP_G_THREADS_SCHEDULER is not set, meaning the 'sexp_applicablep(tmp1)' condition in the 'if' from about line 1015 in vm.c is false, so the scheduler and the usleep never gets called - hence the 100% CPU busy loop on I/O polling.

The easiest way to fix this without touching the code seems to be simply adding an (import (srfi 18)) to your imports.

Better would be to clarify the threading within Chibi-scheme a bit.  What is the SEXP_G_THREADS_SCHEDULER global supposed to be for?  Can it be removed so that SRFI 18 doesn't explicitly need to be imported to get the scheduler to run?

On a related note, in include/chibi/features.h, if one uncomments '#define SEXP_USE_GREEN_THREADS 0', the resulting chibi-scheme REPL core dumps.  If USE_GREEN_THREADS is now baked into the codebase as the only currently working option, perhaps this flag could be removed from the features.h file, and the IFDEF's throughout the codebase simplified to use only the USE_GREEN_THREADS path.  Is there any reason why you would want to turn this off now (i.e. is it supposed to work without them there)?  I can see the point of the option during development, but the green threads implementation appears to be fairly complete now.

**Comment by ashinn** [comment]
cwds is correct.  This also came up in https://groups.google.com/d/msg/chibi-scheme/TbihiLNJsYA/raPt3ne5BwAJ.

The reason the scheduler is not enabled by default is to allow a pluggable
architecture, so you can experiment with smarter schedulers, and also for
systems for which green threads are not yet supported (e.g. windows and
plan9 due to differences in handling non-blocking IO).

I'll take a look to see if I can improve this without falling back to a full
scheduler, maybe just using a longer sleep.

If compiling without threads leads to a segfault that's a regression since
the latest release.  The set of variants in tests/build/build-opts includes:

CPPFLAGS=-DSEXP_USE_GREEN_THREADS=0.

**Comment by cwds** [comment]
Thanks for the explanation on the scheduler, and for the quick integration of the simple syntax fix earlier.  The code, build, and test environment are both easy to follow and work with.

I had not came across the build tests before you mentioned them.  Indeed running them from the Makefile with 'make test-dist' is straightforward once you know what the target does, and most of them pass on the top of the tree on my current environment (OSX 10.11.2, Ubuntu 14.04), including the DSEXP_GREEN_THREADS=0 case.  However, the REPL fails on launching with a segfault when built with the -DSEXP_USE_GREEN_THREADS=0 option, though the R5RS tests from the suite pass on the same binary.  If I get a chance over the next few days, I will take a look to see why this is the case.

**Comment by mnieper** [comment]
Thanks for looking at this issue so promptly! What I don't understand: why is polling at all necessary on a POSIX system (that provides select(2) and poll(2) or even epoll in case of Linux)? I guess, Windows has similar low-level routines for non-blocking I/O.

**Comment by ashinn** [comment]
Because chibi supports non-POSIX systems, and I don't know the Windows or Plan9 equivalents.  I suppose the quickest workaround is to conditionally poll on *nix and busy-wait on the others.

**Comment by cwds** [comment]
About the segfaults, apparently attempting to dynamically link chibi-scheme with SEXP_USE_GREEN_FLAGS=0 with libraries built with SEXP_USE_GREEN_FLAGS=1 winds up in a segfault.  Go figure.  With more care about what is linked to what all those various options indeed work as advertised.

With respect to the I/O and scheduling, given that the stated primary purpose of Chibi is as an extension language for a larger C or similar program, I (as an application writer) probably wouldn't bother attempting to do any significant I/O in scheme; I'd do it in platform-specific C code, exposing an application-specific function to encapsulate whatever high-level feature the I/O was for.  Chibi's current simple and easy to understand implementation looks just fine for it's purpose (say R7RS compliance), though a quick fix so as not to peg the CPU in this case might be nice to have.  

There would seem to be a fair bit of work needed to support a "full" windows version rather than just the psudo-unix MinGW; perhaps if such an exercise were undertaken it would also present a good opportunity to revisit the scheduler.  For instance, building something similar to goroutines would allow for async Scheme code not easily replicated in C.   I suppose you could then build a Go interpreter, should anyone feel the need to do such a thing.

**Comment by mnieper** [comment]
Plan9 seems to support `select` via APE (http://plan9.bell-labs.com/sys/doc/ape.html). I have also found `select` in Windows, but it does not seem to support `stdin`.

So what about the idea of writing a wrapper `select` function that calls the true `select` in case it is available and does polling otherwise?

**Comment by ashinn** [comment]
That should work.  Keep in mind chibi is a one-man project, and a lot of issues are simply things I haven't gotten to yet.

**Comment by ashinn** [comment]
I've pushed an initial change for this.

Ideally Plan9 wouldn't use APE (it's discouraged), but we
don't need it anyway since currently green threads are not
enabled in Plan9.

**Comment by johnwcowan** [comment]
Windows "select" handles sockets only.  There is extremely messy code in Cygwin to work around this problem.

**Comment by mnieper** [comment]
There may be some hope: http://blog.dustinkirkland.com/2016/03/ubuntu-on-windows.html

If they really want to support Ubuntu/Linux binaries without change, they will have to provide a (nearly) complete POSIX environment.

Maybe, the ELF binary of Chibi already runs unmodified under Windows 10.

**Comment by johnwcowan** [comment]
Marc Nieper-Wißkirchen scripsit:

> Maybe, the ELF binary of Chibi already runs unmodified under Windows 10.

I have UfW on my home machine.  I built Chicken, but neither csi nor csc
will run without segfaulting.  I haven't tried Chibi yet.

-- 
John Cowan          http://www.ccil.org/~cowan        cowan@ccil.org
Deshil Holles eamus.  Deshil Holles eamus.  Deshil Holles eamus.
Send us, bright one, light one, Horhorn, quickening, and wombfruit. (3x)
Hoopsa, boyaboy, hoopsa!  Hoopsa, boyaboy, hoopsa!  Hoopsa, boyaboy, hoopsa!
  --Joyce, Ulysses, "Oxen of the Sun"

**Comment by okuoku** [comment]
> I haven't tried Chibi yet.

[I've tried some Scheme implementations](https://github.com/okuoku/yunibase#windows) and Chibi-scheme ran very well there(ie. the only implementation I've ever tried which passes their own test suite). But, IMHO, don't expect too much for Ubuntu on Windows; it's desktop-only optional feature thus it won't support Windows Server and their emulation is not quite accurate yet to do non-trivial workload.

Unfortunately, I/O multiplexing is not something portable on C. [libuv](https://github.com/libuv/libuv) is great on this area and I use it for most of my (proprietary) projects; but I think additional dependency is not preferable on here anyway..

Since threading is now C11 feature, it'd better to add some I/O thread even on green thread to keep portability :P
